File: ios/WatchGuru/Sources/Data/MonthlyBreakdown.swift
import Foundation
import WatchGuruAPI

/// How the monthly chart is named and read aloud.
///
/// Kept out of the view because the spoken summary is the whole chart as far
/// as a VoiceOver user is concerned — it is the one part of a drawing that can
/// be tested, and the one part that would otherwise be written once and never
/// checked again.
enum MonthlyBreakdown {

    /// "Mar 2026" — short, and in the reader's own language.
    static func label(_ bucket: MonthBucket, locale: Locale = .current) -> String {
        var components = DateComponents()
        components.year = bucket.year
        components.month = bucket.month
        guard let date = Calendar(identifier: .gregorian).date(from: components) else {
            return "\(bucket.month)/\(bucket.year)"
        }
        return date.formatted(.dateTime.month(.abbreviated).year().locale(locale))
    }

    /// Every value, in order, for a reader who cannot see the bars.
    ///
    /// A chart owes its readers a table; on a phone this sentence is it.
    static func spokenSummary(_ months: [MonthBucket], locale: Locale = .current) -> String {
        guard !months.isEmpty else {
            return "No viewing recorded in this period."
        }
        let parts = months.map { "\(label($0, locale: locale)), \($0.minutes) minutes" }
        return "Minutes watched by month: " + parts.joined(separator: ", ")
    }

    /// The month with the most watched, or nil when nothing was.
    ///
    /// Nil rather than the first of several zeroes: announcing a "busiest
    /// month" of no hours is a sentence about nothing.
    static func busiest(_ months: [MonthBucket]) -> MonthBucket? {
        months.filter { $0.minutes > 0 }.max { $0.minutes < $1.minutes }
    }
}

File: android/app/src/main/kotlin/dev/dhuelin/watchguru/data/MonthlyBreakdown.kt
package dev.dhuelin.watchguru.data

import dev.dhuelin.watchguru.api.models.MonthBucket
import java.time.Month
import java.time.format.TextStyle
import java.util.Locale

/**
 * How the monthly chart is named and read aloud.
 *
 * Plain Kotlin rather than something inside the composable, because the
 * spoken summary is the whole chart as far as a screen-reader user is
 * concerned -- it is the one part of a drawing that can be tested, and the one
 * part that would otherwise be written once and never checked.
 */
object MonthlyBreakdown {

    /** "Mar 2026" -- short, and in the reader's own language. */
    fun label(bucket: MonthBucket, locale: Locale = Locale.getDefault()): String =
        "${Month.of(bucket.month).getDisplayName(TextStyle.SHORT, locale)} ${bucket.year}"

    /** The single letter under a bar. */
    fun initial(bucket: MonthBucket, locale: Locale = Locale.getDefault()): String =
        Month.of(bucket.month).getDisplayName(TextStyle.NARROW, locale)

    /**
     * Every value, in order, for a reader who cannot see the bars.
     *
     * A chart owes its readers a table; on a phone this sentence is it.
     */
    fun spokenSummary(months: List<MonthBucket>, locale: Locale = Locale.getDefault()): String {
        if (months.isEmpty()) {
            return "No viewing recorded in this period."
        }
        return months.joinToString(
            prefix = "Minutes watched by month: ",
            separator = ", ",
        ) { "${label(it, locale)}, ${it.minutes} minutes" }
    }

    /** The month with the most watched, or null when nothing was. */
    fun busiest(months: List<MonthBucket>): MonthBucket? =
        months.filter { it.minutes > 0 }.maxByOrNull { it.minutes }
}
